simpleviewinc/keg-cli
# Keg switch to args parse

## Context

* Allows running actions for synced repos in a docker container
* Actions should be customized per mutagen.yml config
* Allow calling only a single action from passed in options

## Goal

* Allow passing in `--sync <repo-to-sync>` when starting a docker container
  * This will auto sync that repo
  * This will auto run any actions defined in the `mutagen.yml` config
  * Sync actions should be a object in `yml` format
* Allow only calling a specific action when this format is used => `<repo-to-sync>:<name-of-action>`
  * If no action name is passed, then all actions will be called
* When syncing repos
  * An ENV must be defined in this format => `DOC_<repo-name>_PATH`
  * The ENV value should be the path within the container where the repo should be synced
  * **Example**
	* `DOC_RETHEME_PATH: /keg/keg-components/node_modules/@simpleviewinc/re-theme`
      * This will sync 're-theme' to `/keg/keg-components/node_modules/@simpleviewinc/re-theme` whin the container

## Updates

* `containers/components/mutagen.yml`
  * Added sync actions for syncing `re-theme`
* `containers/components/values.yml`
  * Added notes for ENVS
* `src/libs/docker/container.js`
  * Added more arguments to the `exec` method
* `src/tasks/mutagen/create.js`
  * Added code to run the sync actions
* `src/utils/getters/__tests__/*`
  * Added more tests
* `src/utils/getters/*`
  * Extracted helpers from the `syncService` to be reused
* `src/utils/services/serviceOptions.js`
  * Added detached argument
* `src/utils/services/syncActionService.js`
  * Added service to run sync actions for a container
* `src/utils/services/syncService.js`
  * Cleaned up to work with sync actions

## Testing

* Pull this branch
* Run `yarn install`
* Run `yarn test`
  * Ensure all tests pass

**Test 1**
* Run `keg components start --build --cache false --sync retheme`
  * You only need `--cache false`, if keg components is already built on your machine
  * Check the logs to ensure the sync action for re-theme are run
    * You should see it installing all the dependencies for re-theme
    * You should not see it start re-theme, because this action is run in `detached` mode
* After `keg-components` starts
  * Ensure your terminal is showing the logs for `keg-components`
    * If you disconnected from the logs, run `keg d log components`
      * This for connect back to the logs for the `keg-components` container
  * Modify a file in `re-theme` to cause a re-build
  * Check the terminal for a rebuild of `keg-components`
    * Ensure `keg-components` rebuilds successfully with the `re-theme` changes

**Test 2**
* Kill `keg-components` container => `keg components kill`
* Run `keg sy`, and ensure no syncs are running
* Restart `keg-components` without a sync
  * `keg components start`
* Run the `sync` task for only the `retheme:install` action
  * `keg components sync retheme:install`
  * Ensure the `retheme install` action is run correctly
    * It should create a sync for `re-theme` and install the dependencies, but not run `retheme:start` action
    * Confirm `start` action was not run by modifing a file in your local `re-theme` repo
    * Keg-components should not rebuild
  * Run `keg sy`
    * Ensure a sync was created for `re-theme`
* Run the `sync` task for only the `retheme:start` action
  * Ensure that start action is run correctly
    * You will not see any build logs for re-theme because it's run in detached mode
    * Modify a file in your local `re-theme` repo
    * Keg-components should now rebuild
